cmd/eval.go

```diff
@@ -26,170 +27,121 @@ import (
-					}
-
-					jsonstr = string(file)
-				}
-
-				// Input handler
-				if arguments.input != "" {
-					jsonstr = arguments.input
-				}
-				// Unmarshall json into struct, exit on failure
-				if err := json.Unmarshal([]byte(jsonstr), &snapshotSpec); err != nil {
-					return err
-				}
-
-				if snapshotSpec == nil {
-					return errors.New("ApplicationSnapshot input could not be unmarshalled.")
-				}
+			s, err := determineInputSpec(arguments)
+			if err != nil {
+				return err
 			}
 
+			arguments.spec = s
+
 			return nil
 		},
 		RunE: func(cmd *cobra.Command, args []string) error {
-			validateImage := func(imageRef string) (*policy.Output, error) {
-				out := &policy.Output{}
+			type result struct {
+				err       error
+				component applicationsnapshot.Component
+			}
 
-				i, err := image.NewImageValidator(cmd.Context(), imageRef, arguments.publicKey, arguments.rekorURL)
-				if err != nil {
-					return nil, err
-				}
+			appComponents := arguments.spec.Components
+
+			ch := make(chan result, len(appComponents))
+
+			var lock sync.WaitGroup
+			for _, c := range appComponents {
+				lock.Add(1)
+				go func(comp appstudioshared.ApplicationSnapshotComponent) {
+					defer lock.Done()
+
+					out, err := validate(cmd.Context(), comp.ContainerImage, arguments.policyConfiguration, arguments.publicKey, arguments.rekorURL)
+					res := result{
+						err: err,
+						component: applicationsnapshot.Component{
+							ApplicationSnapshotComponent: appstudioshared.ApplicationSnapshotComponent{
+								Name:           comp.Name,
+								ContainerImage: comp.ContainerImage,
+							},
+							Success: err == nil,
+						},
+					}
 
-				if err := i.ValidateImageSignature(cmd.Context()); err != nil {
-					out.SetImageSignatureCheck(false, err.Error())
-					return nil, err
-				}
-				out.SetImageSignatureCheck(true, "success")
+					// Skip on err to not panic. Error is return on routine completion.
+					if err == nil {
+						res.component.Violations = out.PolicyCheck
+					}
 
-				if err := i.ValidateAttestationSignature(cmd.Context()); err != nil {
-					out.SetAttestationSignatureCheck(false, err.Error())
-					return nil, err
-				}
-				out.SetAttestationSignatureCheck(true, "success")
+					ch <- res
+				}(c)
+			}
 
-				p, err := policy.NewPolicyEvaluator(arguments.policyConfiguration)
-				if err != nil {
-					return nil, err
-				}
+			lock.Wait()
+			close(ch)
 
-				results, err := p.Evaluate(cmd.Context(), i.Attestations())
-				if err != nil {
-					return nil, err
+			components := []applicationsnapshot.Component{}
+			var err error = nil
+			for r := range ch {
+				if r.err != nil {
+					e := fmt.Errorf("error validating image %s of component %s: %w", r.component.ContainerImage, r.component.Name, r.err)
+					err = multierror.Append(err, e)
+				} else {
+					components = append(components, r.component)
 				}
-				out.SetPolicyCheck(results)
-
-				return out, nil
+			}
+			if err != nil {
```

Comment:
FYI, we should change this. We still want to produce a report if there was a validation error that is not due to a policy violation, e.g. missing signature, missing entry in rekor. We have a tracker for this and tackle it later. Just mentioning it here because that will likely do away with this multierror variable.

Reply:
Yeah, we should file a separate story for that, not sure it'll be covered by [HACBS-633](https://issues.redhat.com/browse/HACBS-633). For now I think it is beneficial to report all errors.

Reply:
I believe it will be covered by [HACBS-760](https://issues.redhat.com/browse/HACBS-760).